File: 1406/solution.java
class Solution {
    public String stoneGameIII(int[] stoneValue) {
        int n = stoneValue.length;
        int[] dp = new int[n+1];
        for(int i=n-1;i>=0;i--){
            int sum = 0;
            dp[i] = Integer.MIN_VALUE;
            for(int k=i;k<Math.min(i+3,n);k++){
                sum+=stoneValue[k];
                dp[i] = Math.max(dp[i], sum-dp[k+1]);
            }
        }
        return dp[0]>0 ? "Alice" : dp[0]<0 ? "Bob" : "Tie";
    }
}

File: 1406/solution.py
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        dp = [0]*(n+1)

        for i in range(n-1,-1,-1):
            total = 0
            dp[i] = float('-inf') 
            for k in range(i,min(i+3,n)):
                total+=stoneValue[k]
                dp[i] = max(dp[i], total-dp[k+1])
        
        return "Alice" if dp[0]>0 else "Bob" if dp[0]<0 else "Tie"
